presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java

```diff
@@ -285,4 +513,29 @@ private RowExpression toRowExpression(Expression expression, Rule.Context contex
 
         return SqlToRowExpressionTranslator.translate(expression, types, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), context.getSession(), false);
     }
+
+    private RowExpression toRowExpression(LambdaExpression expression, Rule.Context context, Map<NodeRef<Expression>, Type> lambdaTypes)
+    {
+        return SqlToRowExpressionTranslator.translate(expression, lambdaTypes, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), context.getSession(), false);
+    }
+
+    private RowExpression removeOriginalExpression(RowExpression rowExpression, Rule.Context context)
+    {
+        if (isExpression(rowExpression)) {
+            return toRowExpression(castToExpression(rowExpression), context);
+        }
+        return rowExpression;
+    }
+
+    private RowExpression removeOriginalExpression(RowExpression rowExpression, Rule.Context context, Map<NodeRef<Expression>, Type> lambdaTypes)
+    {
+        if (isExpression(rowExpression)) {
```

Comment:
We have many places in this class doing this check, we should reuse this helper function for all use cases outside aggregation.